### PR TITLE
feat: add ajv

### LINKED ISSUE #309: AJV resolver
**Is your feature request related to a problem? Please describe.**
Build a new AJV resolver for JSON schema

**Describe the solution you'd like**
Having a dedicated AJV resolver

**Describe alternatives you've considered**
Build custom resolver

**Additional context**
We did have an AJV resolver PR (attempted) but failed, wondering if anyone would like to pick it up and give it another try.

https://github.com/react-hook-form/resolvers/pull/59/files

cc @jorisre

### PR DESCRIPTION
**Pull request by jsun969**
~~There are **2 problems** need to be solved~~
1. ~~`mjs` file cannot be created when build. But I have add the build script.~~
2. ~~If field is **required**, user must add `minLength: 1` in the schema. Does this need to be optimized?~~

close #309

### CODE REVIEW

**Comment by getTobiasNielsen** [comment]
> If field is required, user must add minLength: 1 in the schema. Does this need to be optimized?

I'm fairly sure that is how ajv want it to work.

**Comment by bluebill1049** [comment]
This is awesome! @jorisre would be able to help on the first issue since you are an expert on microbundle 😄

**Comment by jsun969** [comment]
According to [Empty values validation in json schema using AJV - Stack Overflow](https://stackoverflow.com/questions/45888524/empty-values-validation-in-json-schema-using-ajv/48778838).

Do I need to add a keyword (maybe `formRequired`) to make the schema more readable?

File: ajv/src/ajv.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+      ...schemaOptions,
+    });
+
+    ajvErrors(ajv);
+
+    // If the field is not require in form, the value is empty (value: "")
+    // But Ajv `required` cannot check empty, it can only check undefined
+    // This keyword is for checking empty, but not available for nest fields
+    // https://stackoverflow.com/a/72374873/15246747
+    ajv.addKeyword({
```
Comment (jsun969) [inline ajv/src/ajv.ts:63]: I've solved the seconed issue. But if the field is nest (like deep object, array, etc.), it may fail.

Comment (jsun969) [reply ajv/src/ajv.ts:63]: Also, this only support string value. I think it it unnecessary to implement this. We can add some tips in the documentation if user meet required error.

**Comment by jsun969** [comment]
> > If field is required, user must add minLength: 1 in the schema. Does this need to be optimized?
> 
> I'm fairly sure that is how ajv want it to work.

🤔 I think we may just ignore the second issue. Just add some tips in documentation. What do you think @bluebill1049

**Comment by bluebill1049** [comment]
> 🤔 I think we can ignore the second issue. Just add some tips in documentation. What do you think? @bluebill1049

thanks, @jsun969 I wouldn't add any "magic" to make it work, and try to stay as native as possible with AJV, at the end of the day people who decided to choose and use AJV resolver should have a clear understanding of its mechanism. By the way, this feature is awesome! so many users have asked for it and it's finally coming into pieces. 

@damien-duignan I may borrow your eyes here as well. 

cc @Moshyfawn

**Comment by jsun969** [comment]
I've fixed the first issue (build mjs error). Could you run the ci again 😄

**Comment by getTobiasNielsen** [comment]
Just to let you know how brilliant the kid is, I pointed out that @jsun969 that none of the other builds produced mjs, and he figured out something had to make the mjs in the code.

https://github.com/react-hook-form/resolvers/blob/af86e823a7d806672ad5a19da7bd03adbef22fec/config/node-13-exports.js#L18-L24

Doesn't this solve the problem?

image

**Comment by bluebill1049** [comment]
Awesome stuff, I will review it soon and get this to release soon! existing stuff.

**Review by bluebill1049 — APPROVED** [review]
Looks amazing to me 🚀  Let's wait for @jorisre to give a review as well.

**Comment by bluebill1049** [comment]
hey @jorisre you are probably busy, I may merge and publish this feature this week. 🙏

**Review by jorisre — APPROVED** [review]

**Comment by jorisre** [comment]
Thanks a lot @jsun969 ! 🚀 👏🏻

**Pull request merged**